src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,131 @@
+    private T[] elementsData;
+    private int capacity;
+    private int size;
+
+    public ArrayList() {
+        this.elementsData = (T[]) new Object[DEFAULT_CAPACITY];
+        this.capacity = DEFAULT_CAPACITY;
+        this.size = START_ARRAY_SIZE;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size() == capacity) {
+            resizeArray();
+        }
+        elementsData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (size() == capacity) {
+            resizeArray();
+        }
+        if (index >= 0 && index <= size) {
+            T[] newElementsData = (T[]) new Object[capacity];
+            for (int i = 0; i < index; i++) {
+                newElementsData[i] = elementsData[i];
+            }
+            newElementsData[index] = value;
+            for (int i = index; i < size(); i++) {
+                newElementsData[i + 1] = elementsData[i];
+            }
+            size++;
+            elementsData = newElementsData;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index doesn't found");
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < size && index >= 0) {
+            return elementsData[index];
+        }
+        throw new ArrayListIndexOutOfBoundsException("Index out of bounds for index: " + index);
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < size && index >= 0) {
+            elementsData[index] = value;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index doesn't found");
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T remove;
+        if (index < size && index >= 0) {
+            T[] newElementsData = (T[]) new Object[capacity];
+            for (int i = 0; i < index; i++) {
+                newElementsData[i] = elementsData[i];
+            }
+            remove = elementsData[index];
+            for (int i = index; i < size(); i++) {
+                newElementsData[i] = elementsData[i + 1];
+            }
+            size--;
+            elementsData = newElementsData;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index doesn't found");
+        }
+        return remove;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int ourIndex = getIndexOfElement(element);
+        if (element == null && elementsData[ourIndex] == null
+                || element != null && elementsData[ourIndex].equals(element)) {
+            return remove(ourIndex);
+        } else {
+            throw new NoSuchElementException("No such element: " + element);
```

Comment:
The NoSuchElementException message should be more informative. Include the element that was not found to provide clearer context in the exception message.